### PR TITLE
Feautre/deploy containers

### PR DESCRIPTION
**Pull request by olegasics**

### CODE REVIEW

File: crates/oct-ctl/Cargo.toml [hunk 1]
```diff
@@ -4,4 +4,10 @@ version = "0.1.0"
...
 [dependencies]
-actix-web = { workspace = true}
\ No newline at end of file
+actix-web = { workspace = true}
+futures = "0.3.31"
+log = "0.4"
+serde = {  workspace = true, features = ["derive"] }
+serde_json = { version = "1.0", workspace = true }
+env_logger = { workspace = true }
+reqwest = { workspace = true }
```
Comment (minev-dev) [inline crates/oct-ctl/Cargo.toml:13]: Here we shouldn't use versions of the dependencies, they are defined in the root `Cargo.toml`

File: docs/oct-deploy-design.md
Comment (minev-dev) [inline docs/oct-deploy-design.md:1]: Looks like it's the old file and it shouldn't be in this PR

File: crates/oct-ctl/src/service.rs
Comment (minev-dev) [inline crates/oct-ctl/src/service.rs:1]: I suggest moving this file to `oct-cli` and name it `oct_ctl_sdk`. With that we won't need to use `oct-ctl` as a dependency of `oct-cli`

File: crates/oct-ctl/src/main.rs [hunk 2]
```diff
@@ -12,17 +14,79 @@ async fn index() -> impl Responder {
...
+            "Success"
+        }
+        Err(err) => {
+            println!("{}", err);
+            "Error"
+        }
+    }
+}
+
+const MAX_SIZE: usize = 262_144; // max payload size is 256k
```
Comment (minev-dev) [inline crates/oct-ctl/src/main.rs:28]: `256 * 1024` is more readable. And if you rename `MAX_SIZE` to `MAX_SIZE_BYTES`, it'll be even more readable

File: Cargo.toml [hunk 3]
```diff
@@ -4,6 +4,7 @@ resolver = "2"
 
 [workspace.dependencies]
 oct-cloud = { path = "crates/oct-cloud" }
+oct-ctl = { path = "crates/oct-ctl" }
```
Comment (minev-dev) [inline Cargo.toml:7]: `oct-ctl` will not be used as a dependency in any crate. At least for now

File: crates/oct-cli/Cargo.toml [hunk 4]
```diff
@@ -5,6 +5,7 @@ edition = "2021"
 
 [dependencies]
 oct-cloud = { workspace = true }
+oct-ctl = { workspace = true }
```
Comment (minev-dev) [inline crates/oct-cli/Cargo.toml:8]: `oct-ctl` will not be used as a dependency in any crate. At least for now

File: crates/oct-cli/src/oct_ctl_sdk.rs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+pub async fn run_container(container_name: String, internal_port: String, external_port: String) -> Result<Response, reqwest::Error> {
+    let client = reqwest::Client::new();
+
+    let mut map = HashMap::new();
+    map.insert("image_uri", container_name.as_str());
+    map.insert("external_port", external_port.as_str());
+    map.insert("internal_port", internal_port.as_str());
+
+    let response = client
+        .post("http://localhost:31888/run-container")
```
Comment (minev-dev) [inline crates/oct-cli/src/oct_ctl_sdk.rs:13]: Pass domain/ip to this function. `localhost` will not work

File: crates/oct-ctl/Cargo.toml [hunk 6]
```diff
@@ -4,4 +4,9 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
-actix-web = { workspace = true}
\ No newline at end of file
+actix-web = { workspace = true}
+serde = {  workspace = true, features = ["derive"] }
+serde_json = { version = "1.0", workspace = true }
```
Comment (minev-dev) [inline crates/oct-ctl/Cargo.toml:9]: Do not pass a version, it's taken from the base Cargo

File: crates/oct-cloud/src/aws.rs [hunk 7]
```diff
@@ -272,8 +272,9 @@ impl Ec2Instance {
     set -e
 
     sudo apt update
-    sudo apt install docker.io -y
-    sudo systemctl start docker
+    sudo apt-get -y install podman
+    sudo systemctl start podman
+    alias docker="podman"
 
     aws ecr get-login-password --region us-west-2 | docker login --username AWS --password-stdin {ecr_repo_uri}
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:279]: Will `docker login` work?

File: crates/oct-ctl/src/main.rs [hunk 8]
```diff
@@ -12,17 +14,79 @@ async fn index() -> impl Responder {
...
+        }
+        Err(err) => {
+            println!("{}", err);
+            "Error"
+        }
+    }
+}
+
+const MAX_SIZE_BYTES: usize = 256 * 1024; // max payload size is 256k
+                                 //
```
Comment (minev-dev) [inline crates/oct-ctl/src/main.rs:29]: Unnecessary comment

File: crates/oct-ctl/src/main.rs [hunk 9]
```diff
@@ -12,17 +14,79 @@ async fn index() -> impl Responder {
...
+        Ok(res) => {
+            println!("Result: {}", String::from_utf8_lossy(&res.stdout));
+            "Success"
+        }
+        Err(err) => {
+            println!("{}", err);
+            "Error"
+        }
+    }
+}
```
Comment (minev-dev) [inline crates/oct-ctl/src/main.rs:26]: Do we still need this endpoint?

File: crates/oct-ctl/src/main.rs [hunk 10]
```diff
@@ -12,17 +14,79 @@ async fn index() -> impl Responder {
...
+    while let Some(chunk) = payload.next().await {
+        let chunk = chunk.unwrap();
+        // limit max size of in-memory payload
+        if (body.len() + chunk.len()) > MAX_SIZE_BYTES {
+            return "Payload too large";
+        }
+        body.extend_from_slice(&chunk);
+    }
+
+    let obj = serde_json::from_slice::<RunContainerPayload>(&body).unwrap();
```
Comment (minev-dev) [inline crates/oct-ctl/src/main.rs:49]: I don't think we need to check the max size now

File: crates/oct-cli/src/main.rs [hunk 11]
```diff
@@ -2,6 +2,7 @@ use std::fs;
 
 use clap::{Parser, Subcommand};
 use log;
+use oct_cli::oct_ctl_sdk::run_container;
```
Comment (minev-dev) [inline crates/oct-cli/src/main.rs:5]: `use crate::oct_ctl_sdk::run_container;`

File: crates/oct-cli/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1 @@
+pub mod oct_ctl_sdk;
```
Comment (minev-dev) [inline crates/oct-cli/src/lib.rs:1]: Move it as `mod oct_ctl_sdk;` (without `pub`) to `crates/oct-cli/src/main.rs` after the imports.
More info https://doc.rust-lang.org/book/ch07-02-defining-modules-to-control-scope-and-privacy.html#modules-cheat-sheet

File: crates/oct-cli/src/main.rs [hunk 13]
```diff
@@ -78,6 +69,23 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                            service.image.to_string(),
+                            service.external_port.to_string(),
+                            service.internal_port.to_string(),
+                            public_ip.to_string(),
+                        )
+                        .await?;
+                    }
+                }
+                None => {
+                    log::info!("Public IP not found");
```
Comment (minev-dev) [inline crates/oct-cli/src/main.rs:86]: `log::error`

File: crates/oct-cloud/src/aws.rs [hunk 14]
```diff
@@ -272,8 +272,8 @@ impl Ec2Instance {
     set -e
 
     sudo apt update
-    sudo apt install docker.io -y
-    sudo systemctl start docker
+    sudo apt-get -y install podman
+    sudo systemctl start podman
 
     aws ecr get-login-password --region us-west-2 | docker login --username AWS --password-stdin {ecr_repo_uri}
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:278]: Change `docker login` to `podman login`

File: crates/oct-ctl/src/main.rs [hunk 15]
```diff
@@ -1,28 +1,60 @@
...
+async fn run(payload: web::Json<RunContainerPayload>) -> impl Responder {
+    let command = Command::new("podman")
+        .args([
+            "run",
+            "-d",
+            "-p",
+            format!(
+                "{external_port}:{internal_port}",
+                external_port = &payload.internal_port,
+                internal_port = &payload.external_port
```
Comment (minev-dev) [inline crates/oct-ctl/src/main.rs:22]:
```
internal_port = &payload.internal_port,
external_port = &payload.external_port
```

File: crates/oct-cli/src/oct_ctl_sdk.rs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+            "http://{public_ip}:31888/run-container",
+            public_ip = public_ip
+        ))
+        .header("Content-Type", "application/json")
+        .header("Accept", "application/json")
+        .body(serde_json::to_string(&map).unwrap())
+        .send()
+        .await;
+
+    return response;
```
Comment (minev-dev) [inline crates/oct-cli/src/oct_ctl_sdk.rs:28]: You can do just
```
client
    .post(format!(
        "http://{public_ip}:31888/run-container",
        public_ip = public_ip
    ))
    .header("Content-Type", "application/json")
    .header("Accept", "application/json")
    .body(serde_json::to_string(&map).unwrap())
    .send()
    .await
```

**Review by minev-dev — APPROVED** [review]

**Review by andagaev — APPROVED** [review]

**Pull request merged**